### ISSUE TITLE
Bug: [switch-exhaustiveness-check] incorrectly flags enum switches as non-exhaustive

### ISSUE THREAD

**Issue by maRci002** [typescript-eslint/typescript-eslint#10524]
### Before You File a Bug Report Please Confirm You Have Done The Following...

- [X] I have tried restarting my IDE and the issue persists.
- [X] I have updated to the latest version of the packages.
- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22package%3A+eslint-plugin%22) and found none that matched my issue.
- [X] I have [read the FAQ](https://typescript-eslint.io/linting/troubleshooting) and my problem is not listed.

### Playground Link

https://typescript-eslint.io/play/#ts=5.7.2&fileType=.tsx&code=KYOwrgtgBAYg9nKBvKAhAhgJwDRvQLygF8BYAKHIDMwQBjAFwEs4Qp7gBnegCkoQC5YCAJSCAbnEYATZOShQOAd0b1aACyi8RssvPm10HYELgA6DPn5y98gEaZg6ANYBua-KnBK6MABt6Vro2UPaOrtakZJFAA&eslintrc=N4KABGBEBOCuA2BTAzpAXGUEKQAIBcBPABxQGNoBLY-AWhXkoDt8B6ZAd0vzIAt6AHrwCGsZPkoA3RExTJafRGQDW6KImjQA9tEjgwAXxAGgA&tsconfig=N4KABGBEDGD2C2AHAlgGwKYCcDyiAuysAdgM6QBcYoEEkJemy0eAcgK6qoDCAFutAGsylBm3TgwAXxCSgA&tokens=false

### Repro Code

```TypeScript
enum Foo { Bar, Baz }

function test(foo: Foo): void {
    switch (foo) {
        case Foo.Baz:
            break;
        default:
            break;
    }
}
```

### ESLint Config

```javascript
module.exports = {
  parser: "@typescript-eslint/parser",
  rules: {
    "@typescript-eslint/switch-exhaustiveness-check": "error"
  },
};
```

### tsconfig

```jsonc
{
  "compilerOptions": {
    "strictNullChecks": true
  }
}
```

### Expected Result

No linting error when all cases of the `Foo` enum are correctly covered by the `switch` statement, including the `default` case.

### Actual Result

The rule `@typescript-eslint/switch-exhaustiveness-check` incorrectly reports that the `switch` statement is not exhaustive, even when all enum cases are covered explicitly or via the `default` case.

Error message:  
`Switch is not exhaustive. Cases not matched: Foo.Bar`

### Additional Info

The issue did not occur in `@typescript-eslint/eslint-plugin` version `8.12.0`. The problem started appearing in newer versions.

**Comment by kirkwaiblinger** [comment]
The default behavior was changed, yeah. See https://typescript-eslint.io/rules/switch-exhaustiveness-check/#considerdefaultexhaustiveforunions and the release notes for the 8.12.0 (and its patches), https://github.com/typescript-eslint/typescript-eslint/releases

**Comment by maRci002** [comment]
Thank you so much @kirkwaiblinger
